src/cargo/sources/registry/mod.rs

```diff
@@ -709,26 +709,33 @@ impl<'cfg> Source for RegistrySource<'cfg> {
         kind: QueryKind,
         f: &mut dyn FnMut(Summary),
     ) -> Poll<CargoResult<()>> {
-        // If this is a precise dependency, then it came from a lock file and in
+        let mut req = dep.version_req().clone();
+
+        // Handle `cargo update --precise` here.
+        if let Some((_, requested)) = self
+            .source_id
+            .precise_registry_version(dep.package_name().as_str())
+            .filter(|(c, _)| req.matches(c))
+        {
+            req.lock_to(&requested);
```

Comment:
`lock_to` was changed have the same semantics as `--precise`, but it's possible we should have two different kind of locked requirements.